### ISSUE TITLE
Improved revocation checker

### ISSUE THREAD

**Issue by HylkeB** [appmattus/certificatetransparency#52]
I have experimented with an improved revocation checking mechanism, using built-in android tooling, that works from api level 24 and up. For now I have only published it as a gist, and I am using it myself in my project.

If the community can do a quick review of the solution provided here https://gist.github.com/HylkeB/f9f3c26398c49985e54545df1b76b062 and provide feedback on it, I think it might be a good idea to replace the certificate revocation implementation currently in this library for my one. If others agree I might be able to find the time to create a PR for this.

The downside on the current certificate revocation implementation in this library is that it requires an app update, basically creating another source of truth for the revocation status of a certificate. WebPKI already has solutions for revocation, using OCSP, CRL and OCSPStapling, so it would be wise to reuse the existing solutions for that, and not create a new one.

With my solution the revocation interceptor uses the built-in `CertPathValidator` and `PKIXRevocationChecker` to check if none of the used certificates in the certificate chain are revoked (which in turn uses the certificate to find the OCSP responder url, and as a fallback uses the CRL url to determine if the certificate is revoked or not). On older android versions, this is disabled by default; or configured as a soft-fail. My solution requires with certainty that the certificate is not revoked. If it is not possible to check the OCSP status and CRL (verified by going to airplane mode just before the check using breakpoints) it will block the connection.

**Comment by mattmook** [comment]
This does look very interesting; ultimately revocation is not core functionality of the library and is not something I want in it long term; my implementation was certainly always a stop-gap!

i.e. it should live in an independent certificate-revocation library.

Worth reading this article https://scotthelme.co.uk/revocation-is-broken/ and verifying your solution against that as there are issues relying on some of those forms of revocation.

I'd previously created a note to myself about revocation and I'm adding a link to your solution too for when I have time to follow up.

**Comment by HylkeB** [comment]
Hi Matt, thanks for responding.

I have to disagree that revocation checking is not a core functionality. In my opinion revocation checking should be a core functionality of this library if you want to promote this library as a replacement to certificate pinning (as stated in the readme).

If you have certificate transparency enforced in your http client as a replacement for certificate pinning, what you are saying is that you rely on your monitoring infrastructure to be informed of maliciously issued certificates and then having the CA's revoke these maliciously issued certificates. 

This leaves a small window of attack that you are fine with, as it reduces the LCM effort of constantly keeping the certificate pin in sync with the server certificate. The small window of attack is the time it takes from when the attacker was able to get an maliciously issued certificate, and has it published to the logs, until your monitoring infrastructure has successfully revoked that malicious certificate.

Now if you say that the certificate revocation checks are not a core part of this library, then even if you have the CT monitoring infrastructure up and running, and you get malicious certificates revoked in reasonable time, that does not matter for the attacker since the app you're trying to protect from MITM attacks thinks its all ok. The malicious certificate is valid, its signed by a valid CA, its embedded in the CT logs, its also revoked but thats not checked, so its not a proper MITM protection mechanism without revocation checks.


Also; I've read the article you linked, in the article they make some good points; but thats all from a browser standpoint. A browser does not have to replace certificate pinning to protect against MITM attacks. This library is for mobile apps that want to make sure they connect with their own server, and not a MITM, and they probably want to replace certificate pinning.

Their point about soft fails vs hardfails: since we're writing something proprietary, we can just hard fail when the OCSP or CRL is not available. (thats what I do in my gist as well)

About the availability: we could introduce some caching mechanism, it tries to verify the revocation status every time, cache the result, and keep that in a cache for x amount of time, which increases both the reliability (yay) and the attack window (aww).

**Comment by mattmook** [comment]
Worth noting that browsers don't do certificate pinning anymore!

This libraries sole purpose is for performing CT checks not revocation checks.

Thats not to say revocation checks aren't important; they are for the reasons you state. I'm not happy with that small window for attack either.

When I've done talks on the topic I've made sure I've stated that as well, it is one of CTs bigger flaws - it's not just about monitoring.

So this isn't to say that we shouldn't do them I'd just rather the code lived in a separate library so both libraries have a single purpose. The scope for a revocation library can be huge, from managing CRL lists to OCSP etc, or even the ability to add your own manual revocations (for those times you know you have a bad cert perhaps)
